### ISSUE TITLE
Flag register updated too early

### ISSUE THREAD

**Issue by MalteT** [klemens/2i-emulator#7]
Emulating the following program:
```
00000: 00 00001 00 000 0111 00 0 1 0101 1 | ADDS 7, 1, CHFLGS
00001: 00 00010 00 000 1111 00 0 1 0101 0 | ADDS 0xFF, 1, CHFLGS
00010: 00 00011 00 000 0000 00 0 0 0000 0 | NOP
00011: 00 00100 00 000 0000 00 0 0 0000 1 | NOP, CHFLGS
00100: 00 00101 00 000 0000 00 0 0 0000 0 | NOP
00101: 00 00000 00 000 0000 00 0 0 0000 0 | JMP 0
```

The flags are set according to the following table:

| CLK 	| ADDR 	| Real machine   | Emulator           	|
| ----:	| ------	| ----------------	   | --------------------	|
|   0 	| 0    	| CO             	| (ZO, CO, NO undef) 	|
|   1 	| 1    	| ZO, CF         	| CO, CF             	|
|   2 	| 2    	| CO, ZO, CF     	| ZO, CF             	|
|   3 	| 3    	| CO, ZO, CF     	| CO, ZO, CF         	|
|   4 	| 4    	| CO, ZO, CF, ZF 	| CO, ZO, CF, ZF     	|
|   5 	| 5    	| CO, ZO, CF, ZF 	| CO, ZO, CF, ZF     	|
|   6 	| 0    	| CO, CF, ZF     	| CO, ZO, CF, ZF     	|
|   7 	| 1    	| ZO, CF         	| CO, CF             	|

CO, ZO, NO seem to be one clock signal behind.

**Comment by klemens** [comment]
I just tested the program (instruction 1 doesn't CHFL) and the flags seem correct to me?

|instruction|flag/register after executing the instruction|
|-|-|
|TEST R0 + 7 + 1 (ADDS); CHFL|`C: 1/1 N: 0/0 Z: 0/0`|
|TEST R0 + FF + 1 (ADDS)|`C: 0/1 N: 0/0 Z: 1/0`|
|NOP (ADDH)|`C: 1/1 N: 0/0 Z: 1/0`|
|NOP (ADDH); CHFL|`C: 1/1 N: 0/0 Z: 1/1`|
|NOP (ADDH)|`C: 1/1 N: 0/0 Z: 1/1`|
|NOP (ADDH); JMP 00000|`C: 1/1 N: 0/0 Z: 1/1`|

Edit: Sorry for submitting too early, hit the enter key a little too hard…

**Comment by MalteT** [comment]
I tested this on the real machine just now, and the flag register of the emulator seems to adjust it's values one clock to early.

The results I got were:

| instruction                  	| flags/register       	|
|------------------------------	|----------------------	|
| TEST R0 + 7 + 1 (ADDS); CHFL 	| `C: 1/0 N: 0/0 Z: 0/0` 	|
| TEST R0 + FF + 1 (ADDS)      	| `C: 0/1 N: 0/0 Z: 1/0` 	|
| NOP (ADDH)                   	| `C: 1/1 N: 0/0 Z: 1/0` 	|
| NOP (ADDH); CHFL             	| `C: 1/1 N: 0/0 Z: 1/0` 	|
| NOP (ADDH)                   	| `C: 1/1 N: 0/0 Z: 1/1` 	|
| NOP (ADDH); JMP 00000        	| `C: 1/1 N: 0/0 Z: 1/1` 	|

Edit: Format

**Comment by klemens** [comment]
Ok, so at least the normal flags are correct. I haven't used the flags register much myself, so it is very much possible that there is something wrong.

Could you execute the following program on the real machine and note the execution sequence?

```
00000: 00 00001 00 000 0111 00 0 1 0101 1 # TEST R0 + 7 + 1 (ADDS); CHFL
00001: 01 00011 00 000 0000 00 0 0 0011 0 # NOP; JMP 0001{CF}
00010: 00 00100 00 000 0000 00 0 0 0011 0 # should NOT be taken
00011: 00 00100 00 000 0000 00 0 0 0011 0 # SHOULD be taken
00100: 01 00111 00 000 1111 00 0 1 0101 1 # TEST R0 + 0xFF + 1 (ADDS); JMP 0011{CF}; CHFL
00101: 00 00101 00 000 0000 00 0 0 0011 0 # END (LOOP)
00110: 00 00101 00 000 0000 00 0 0 0011 0 # should NOT be taken
00111: 00 00101 00 000 0000 00 0 0 0011 0 # SHOULD be taken
```

The emulator executes this as `0, 1, 3, 4, 6, 5, 5, …`, but I have the strong feeling that it should rather go though 7 than 6, i.e. first calculate the next address and then update the flags register.

**Comment by MalteT** [comment]
Yes, but probably not before Tuesday.

I should also mention, that I didn't notice it, but a student whose program executed differently on the real machine. I don't remember his name and he didn't seem eager to get it fixed..

**Comment by MalteT** [comment]
I got the following results: `0, 1, 3, 4, 7, 5, 5, 5...`
So it behaves as you predicted..

**Comment by klemens** [comment]
Fixed in 2d6f0b727faed7ef8c35da43ff6b7e61198af89e and released as [version 2.0.2](https://github.com/klemens/2i-emulator/releases/tag/v2.0.2). Thanks for taking the time to report the issue!
